### PR TITLE
Port test_introducer.py to Python 3

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3514

### CODE REVIEW

**Review by exarkun — CHANGES_REQUESTED** [review]
Thanks.  Looks pretty good.  Some comments inline.  Here's N more:

* Do we have a best practice for strings that represent paths?  The first two tests in
 `test_introducer` make two different choices - native string literal and unicode string literal.
* in `test_introducer` there is a `TooNewServer` that still uses native strings for its `VERSION`.  It's used by a test that asserts some error has happened.  I guess that test is trying to assert that the right error happened but since the real `IntroducerService` has had its `VERSION` made into byte strings instead of native strings I wonder if that test is really working right.

File: src/allmydata/test/test_introducer.py [hunk 1]
```diff
@@ -487,7 +500,7 @@ def got(key_s_or_tubid, ann, announcements):
             subscribing_clients.append(c)
             expected_announcements[i] += 1 # all expect a 'storage' announcement
 
-            node_furl = tub.registerReference(Referenceable())
+            node_furl = tub.registerReference(Referenceable()).encode("utf-8")
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/test/test_introducer.py:503]: Yeah I am wondering if furl as bytes was a mistake.

[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/test/test_introducer.py:503]: Although that may be due to Foolscap intake? I forget. Will maybe do followup issue.

File: src/allmydata/introducer/client.py [hunk 2]
```diff
@@ -114,19 +117,23 @@ def _save_announcements(self):
         announcements = []
         for _, value in self._inbound_announcements.items():
             ann, key_s, time_stamp = value
+            # On Python 2, bytes are stored as Unicode. To minimize changes, Python
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/introducer/client.py:120]: I don't understand the first sentence either. Hm...

[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/introducer/client.py:120]: Ah, updated as follows:

```
# On Python 2, bytes strings are encoded into YAML Unicode strings.
# On Python 3, bytes are encoded as YAML bytes. To minimize
# changes, Python 3 for now ensures the same is true.
```

File: src/allmydata/test/common.py [hunk 3]
```diff
@@ -1204,7 +1206,9 @@ class AsyncBrokenTestCase(_TestCaseMixin, TestCase):
     pass with ``AsyncTestCase``.
     """
     run_tests_with = EliotLoggedRunTest.make_factory(
-        AsynchronousDeferredRunTestForBrokenTwisted.make_factory(timeout=60.0),
+        AsynchronousDeferredRunTestForBrokenTwisted.make_factory(
+            timeout=60.0, suppress_twisted_logging=False,
+            store_twisted_logs=False),
     )
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/test/common.py:1212]: https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3543 opened, will delete from this branch.

File: src/allmydata/test/test_introducer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/test/test_introducer.py:503]: https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3544#ticket

**Comment by itamarst** [comment]
I could not reproduce the Python 3 failure locally, but tried to fix it based on logs. Merged forward, hopefully correctly.

File: src/allmydata/introducer/client.py [hunk 4]
```diff
@@ -35,11 +38,10 @@ def __init__(self, tub, introducer_furl,
         self._sequencer = sequencer
         self._cache_filepath = cache_filepath
 
-        self._my_subscriber_info = { "version": 0,
-                                     "nickname": self._nickname,
-                                     "app-versions": [],
```
Comment (exarkun) [inline src/allmydata/introducer/client.py:40]: I'm guessing you had a merge conflict to resolve here.  I think `app-versions` should remain in this dict with an empty list as its value.  This was a consequence of https://github.com/tahoe-lafs/tahoe-lafs/pull/907.

**Review by exarkun — APPROVED** [review]
Thanks.  Looks good to me.  Just one minor comment inline.  Please address and then merge.

File: src/allmydata/introducer/client.py (hunk 4 above)
Comment (itamarst) [reply src/allmydata/introducer/client.py:40]: Yeah, must've done it wrong.

**Pull request merged**